### PR TITLE
Reader Caterpillars: add pagination

### PR DESCRIPTION
**Pull request by samouri**
As of the time of this writing, Caterpillars only have support for dealing with the latest 50 comments. This PR can be closed when they can handle all the comments.

This was implemented by adding a new concept to redux concepts: `contiguous`.

A comment is considered contiguous if it was loaded as part of a block of comments as opposed to as an individual comment.  That assists with knowing how to paginate.

### CODE REVIEW

File: client/blocks/comments/post-comment.jsx [hunk 1]
```diff
@@ -129,7 +129,7 @@ class PostComment extends React.PureComponent {
 
 		const immediateChildren = get( commentsTree, [ id, 'children' ], [] );
 		return immediateChildren.concat(
-			flatMap( immediateChildren, child => this.getAllChildrenIds( child.ID ) )
```
Comment (samouri) [inline client/blocks/comments/post-comment.jsx]: this is a bug in current code

**Comment by samouri** [comment]
I'd want to touch up code a bit, but functionality is ready for review here.

**Comment by jancavan** [comment]
I think the comment count is incorrect. In the example below, it says "Kjell and 4 more commented", but when you expand the caterpillar, it's really only hiding a single comment and that's Ballio's. It should say "Ballio Chan commented":

https://cloudup.com/cNkYo70yAdp

It's working much better in this branch though than on wpcalypso, which initially shows 2nd-level comments then shows top-level ones after clicking the caterpillar:

https://cloudup.com/c0lyPB8XKtf

I'm unable to get the video to convent on Cloudup so it should be downloaded first to view to it.

**Comment by samouri** [comment]
rebased off the latest perf improvements and will look into the count bug

**Comment by samouri** [comment]
just need to fix a snapshot :)

**Comment by samouri** [comment]
Created a slight spinoff where i test out using the highest up parent as the [context-comment](https://github.com/Automattic/wp-calypso/issues/16933) as opposed to the immediate parent: https://github.com/Automattic/wp-calypso/pull/18499

**Comment by fraying** [comment]
Comment count bug fixed!

File: client/blocks/conversations/list.jsx [hunk 2]
```diff
@@ -94,34 +95,102 @@ export class ConversationCommentList extends React.Component {
...
+		get( commentsTree, this.getParentId( commentsTree, childId ), false );
+
+	getInaccessibleParentsIds = ( commentsTree, commentIds ) => {
+		// base case
+		if ( size( commentIds ) === 0 ) {
+			return [];
+		}
+
+		const withParents = filter( commentIds, id => this.commentHasParent( commentsTree, id ) );
+		const accessible = map(
```
Comment (blowery) [inline client/blocks/conversations/list.jsx]: could use `partition` here and save a walk?

File: client/blocks/conversations/list.jsx [hunk 3]
```diff
@@ -94,34 +95,102 @@ export class ConversationCommentList extends React.Component {
...
+		);
+		const inaccessible = map(
+			filter( withParents, id => ! this.commentParentIsLoaded( commentsTree, id ) ),
+			id => this.getParentId( commentsTree, id )
+		);
+
+		return inaccessible.concat( this.getInaccessibleParentsIds( commentsTree, accessible ) );
+	};
+
+	getCommentsToShow = () => {
```
Comment (blowery) [inline client/blocks/conversations/list.jsx]: wonder if we should do this work as part of willMount / willReceiveProps? doing it every render might get spendy? probably nothing to worry about now though.

File: client/state/comments/reducer.js [hunk 4]
```diff
@@ -111,7 +111,11 @@ export function items( state = {}, action ) {
 				[ stateKey ]: map( state[ stateKey ], updateComment( commentId, comment ) ),
 			};
 		case COMMENTS_RECEIVE:
-			const { skipSort, comments } = action;
+			const { skipSort } = action;
+			const comments = map( action.comments, cmt => ( {
```
Comment (blowery) [inline client/state/comments/reducer.js]: `comment` over `cmt`?

**Comment by blowery** [comment]
Tested out well, even with posts with 200+ comments. Nice work!

File: client/blocks/conversations/list.jsx (hunk 2 above)
Comment (samouri) [reply client/blocks/conversations/list.jsx]: <3 partition. I'll test it out

File: client/state/comments/reducer.js (hunk 4 above)
Comment (samouri) [reply client/state/comments/reducer.js]: Hmm.  I think i remember there being a name conflict. I'll fix

Comment (samouri) [reply client/state/comments/reducer.js]: yep. theres a name conflict.  I'll make it `_comment`

**Pull request merged**